gluten-data/src/main/scala/io/glutenproject/vectorized/ColumnarBatchSerializer.scala

```diff
@@ -198,18 +206,23 @@ private class ColumnarBatchSerializerInstance(
 
       override def close(): Unit = {
         if (!isClosed) {
-          if (numBatchesTotal > 0) {
-            readBatchNumRows.set(numRowsTotal.toDouble / numBatchesTotal)
-          }
-          numOutputRows += numRowsTotal
-          wrappedOut.close()
-          byteIn.close()
-          if (cb != null) {
-            cb.close()
-          }
+          TaskResources.removeResource(resourceId)
```

Comment:
So the call to `#removeResource` may happen inside a call to `TaskResourceRegistry#releaseAll`. Should we double-check whether the code brings inconsistency?

To simplify the procedure, I'd suggest returning a `TaskResourceHandle[T]` (or something like that) from TaskResource's `addResource...` `addRecycler...` APIs, then caller could keep the handle and manually release the resource and remove it from registry to make it GC-able by calling `TaskResourceHandle#release` API when needed. Then we don't need to add a new property `ID` to task resource definition.

Reply:
> So the call to `#removeResource` may happen inside a call to `TaskResourceRegistry#releaseAll`. Should we double-check whether the code brings inconsistency?

As `TaskResourceRegistry#releaseAll` is called at the end of taks, should calling 'close` explicitly always be earlier than the former one? Based on my understanding, calling `close` after `TaskResourceRegistry#releaseAll` should be considered as invalid code path.

Reply:
I see. It's possible that releasing other `TaskResource` from `TaskResourceRegistry#releaseAll` calls this `close` function.

Reply:
Sorry I got it wrong. You are right that it's not possible to be inside another operation on `TaskResourceRegistry` since you are calling `close0` rather than `close` in the recycler.